robot/packages/local_package_manager.go

```diff
@@ -0,0 +1,250 @@
+// local + cloud manager need separate parent dirs so they don't delete each other in Cleanup.
+func LocalPackagesDir(packagesDir string) string {
+	return filepath.Clean(packagesDir) + "-local"
+}
+
+// PackagePath returns the package if it exists and already downloaded. If it does not exist it returns a ErrPackageMissing error.
+func (m *localManager) PackagePath(name PackageName) (string, error) {
+	return string(name), nil
+}
+
+// Close manager.
+func (m *localManager) Close(ctx context.Context) error {
+	return nil
+}
+
+// fileCopyHelper is the downloadCallback for local tarball modules.
+func (m *localManager) fileCopyHelper(ctx context.Context, path, dstPath string) (string, error) {
+	src, err := os.Open(path) //nolint:gosec
+	if err != nil {
+		return "", err
+	}
+	defer src.Close()              //nolint:errcheck
+	dst, err := os.Create(dstPath) //nolint:gosec
+	if err != nil {
+		return "", err
+	}
+	defer dst.Close() //nolint:errcheck
+	nBytes, err := io.Copy(dst, src)
+	if err != nil {
+		return "", err
+	}
+	m.logger.Debugf("copied %d bytes to %s", nBytes, dstPath)
+	// note: we can hardcode expected contentType because this is probably a synthetic package which already passed tarballExtensionsRegexp
+	return allowedContentType, nil
+}
+
+// getAddedAndChanged is a helper for managing maps of things. It returns (map of existing, slice of added).
+func getAddedAndChanged[Key comparable, ManagedVal, Val any](previous map[Key]ManagedVal, incoming []Val, keyFn func(Val) Key,
+	compareFn func(ManagedVal, Val) bool,
+) (map[Key]ManagedVal, []Val) {
+	existing := make(map[Key]ManagedVal, len(previous))
+	changed := make([]Val, 0)
+	for _, val := range incoming {
+		key := keyFn(val)
+		if oldVal, ok := previous[key]; ok {
+			if compareFn(oldVal, val) {
+				existing[key] = oldVal
+				continue
+			}
+		}
+		changed = append(changed, val)
+	}
+	return existing, changed
+}
+
+// getAddedAndChanged specializes the generic function for managedModuleMap.
+func (m managedModuleMap) getAddedAndChanged(incoming []config.Module) (managedModuleMap, []config.Module) {
+	return getAddedAndChanged(m, incoming,
+		func(mod config.Module) string { return mod.Name },
+		func(old *managedModule, incoming config.Module) bool { return old.module.ExePath == incoming.ExePath },
+	)
+}
+
+// Sync for the localManager manages copying of local tarballs.
+func (m *localManager) Sync(ctx context.Context, packages []config.PackageConfig, modules []config.Module) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// overwrite incoming modules with filtered slice; we only manage local tarball modules
+	modules = rUtils.FilterSlice(modules, config.Module.NeedsSyntheticPackage)
+	existing, changed := m.managedModules.getAddedAndChanged(modules)
+	if len(changed) > 0 {
+		m.logger.Info("Local package changes have been detected, starting sync")
+	}
+
+	start := time.Now()
+	var outErr error
+	for idx, mod := range changed {
+		pkgStart := time.Now()
```

Comment:
we should be able to just call `SyncOne` inside here, right?

Reply:
nuanced and weird: with local tarballs, we don't always want the 'copy if newer' behavior

`r.localPackages.Sync` gets called at the top of `localRobot.Reconfigure` before the diff -- that means it operates on all local tarball modules, any time there's a config change in the config watcher in `robotServer.serveWeb`

(even an unrelated config change can trigger it)

two problems with that:

1. user may not necessarily want file changes to trigger a process restart, especially because this will appear to be triggered by an unrelated config change, which is confusing
2. because local tarballs are unversioned, they rely on the `[stop, replace, start]` order; if we had SyncOne-type logic at the top of localRobot.Reconfigure, the localManager would be deleting the cwd of the running module while it's running

if we come up with an architecture solution for 'persistent module state', problem 2 goes away but problem 1 remains.